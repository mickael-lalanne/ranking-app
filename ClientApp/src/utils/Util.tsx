import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
