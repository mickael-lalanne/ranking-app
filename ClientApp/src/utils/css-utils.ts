export const ELEMENT_SIZE: string = '75px';
