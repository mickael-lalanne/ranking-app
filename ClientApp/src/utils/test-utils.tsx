import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
