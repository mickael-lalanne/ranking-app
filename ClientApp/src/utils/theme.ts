import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
