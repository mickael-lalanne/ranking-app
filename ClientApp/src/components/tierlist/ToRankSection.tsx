import React, { useState, useEffect } from 'react';
import { css } from '@emotion/css';
import { Template } from '../../models/Template';
import { Element, RankedElement } from '../../models/Element';
import ElementPreview from '../shared/ElementPreview';
import { useDrop } from 'react-dnd';
import { ETierlistDragItem } from '../../models/Tierlist';
import { ELEMENT_SIZE } from '../../utils/css-utils';

const ToRankSection = ({ template, rankedElements, unrankHandler }: {
    template?: Template,
    rankedElements: RankedElement[],
    unrankHandler: (elementId: string) => void
}) => {
    const [notRankedElements, setNotRankedElements] = useState<Element[]>([]);

    // If user drop a ranked element in the "to rank section", unrank it
    const [{ isOver, canDrop }, drop] = useDrop(
        () => ({
            accept: ETierlistDragItem.Element,
            drop: (item: Element) => {
                unrankHandler(item.id!);
            },
            canDrop: (item: Element) => {
                // Only the ranked elements can be dropped in the unrank area
                return !!rankedElements.find(elt => elt.elementId === item.id);
            },
            collect: monitor => ({
                isOver: !!monitor.isOver(),
                canDrop: monitor.canDrop()
            }),
        }),
        [notRankedElements, rankedElements]
    );

    /**
     * Called when the selected template has changed or when a element has been ranked
     * Determine all not ranked elements
     */
    useEffect(() => {
        if (template) {
            const notRankedElements = template.elements.filter(
                templateElt => rankedElements.findIndex(rankedElt => rankedElt.elementId === templateElt.id) === -1
            );
            setNotRankedElements(notRankedElements);
        }
    }, [template, rankedElements]);

    const ElementsList = (): React.JSX.Element[] => {
        const list: React.JSX.Element[] = [];

        notRankedElements.forEach(element => {
            list.push(
                <div key={element.id} style={{ margin: ELEMENTS_MARGIN + 'px' }}>
                    <ElementPreview element={element}/>
                </div>
            );
        });

        return list;
    };

    const UnrankDropZone = (): React.JSX.Element | false => {
        return canDrop && <div
            style={{ backgroundColor: isOver ? 'yellow' : 'white'}}
            className={unrank_zone_style}
        >
            Unrank element
        </div>;
    };

    return(<>
        <div ref={drop} className={section_container_style}>
            {UnrankDropZone()}
            {ElementsList()}
        </div>
    </>);
};

export default ToRankSection;

/**
 * CSS STYLES
 */
const ELEMENTS_MARGIN: number = 2;
const section_container_style = css({
    position: 'relative',
    display: 'flex',
    justifyContent: 'flex-start',
    margin: '20px 0',
    flexWrap: 'wrap',
    minHeight: parseFloat(ELEMENT_SIZE) + ELEMENTS_MARGIN * 2 + 'px',
    boxShadow: 'rgba(0, 0, 0, 0.12) 0px 1px 3px, rgba(0, 0, 0, 0.24) 0px 1px 2px',
});

const unrank_zone_style = css({
    position: 'absolute',
    top: 0,
    left: 0,
    height: '100%',
    width: '100%',
    zIndex: 1,
    opacity: 0.80,
    border: '2px solid black',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
});
