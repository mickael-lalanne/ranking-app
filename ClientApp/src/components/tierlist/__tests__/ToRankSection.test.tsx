import * as React from 'react';
import { renderWithProviders } from '../../../utils/test-utils';
import ToRankSection, { ToRankSectionProps } from '../ToRankSection';

const tierId: string = 'my_tier_id';
const elementId: string = 'my_element_id';
const position: number = 2;

const DEFAULT_PROPS: ToRankSectionProps = {
    rankedElements: [{ tierId, elementId, position }],
    template: {
        id: 'my_template_id',
        name: 'template name',
        tiers: [{ id: tierId, name: 'tier name', rank: 1 }],
        elements: [{ id: elementId, name: 'elt name', image: 'elt image' }],
    },
    unrankHandler: jest.fn(),
};

const ToRankSectionTest = (props: ToRankSectionProps) => {
    return (
        <ToRankSection
            rankedElements={props.rankedElements}
            template={props.template}
            unrankHandler={props.unrankHandler}
        />
    );
};

describe('Screenshot', () => {
    it('takes empty section screenshot', () => {
        const component = renderWithProviders(ToRankSectionTest(DEFAULT_PROPS));

        expect(component.baseElement).toMatchSnapshot();
    });

    it('takes section with unrank elements screenshot', () => {
        const component = renderWithProviders(
            ToRankSectionTest({ ...DEFAULT_PROPS, rankedElements: [] })
        );

        expect(component.baseElement).toMatchSnapshot();
    });
});

// FYI: d&d behaviors are tested in cypress with e2e tests
