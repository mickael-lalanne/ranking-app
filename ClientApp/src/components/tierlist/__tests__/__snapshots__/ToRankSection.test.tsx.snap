// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes empty section screenshot 1`] = `
<body>
  <div>
    <div
      class="css-1njjqgc"
      data-cy="to-rank-section"
      data-testid="to-rank-section"
    />
  </div>
</body>
`;

exports[`Screenshot takes section with unrank elements screenshot 1`] = `
<body>
  <div>
    <div
      class="css-1njjqgc"
      data-cy="to-rank-section"
      data-testid="to-rank-section"
    >
      <div
        style="margin: 2px;"
      >
        <div
          class="css-13ambje"
          data-cy="element-preview"
          data-testid="element-preview"
          draggable="true"
          style="height: 75px; width: 75px; opacity: 1;"
        >
          <img
            class="css-40zi5q"
            data-cy="element-preview-img"
            src="https://res.cloudinary.com/ranking-app/image/upload/elt image"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
